### PR TITLE
feat: `avm/res/container-service/managed-cluster` Update parameter's description which is in preview feature.

### PR DESCRIPTION
**Pull request by zedy-wj**
## Description

Fixes https://github.com/Azure/bicep-registry-modules/pull/3104#issuecomment-2438488253

## Pipeline Reference

| Pipeline |
| -------- |
|[avm.res.container-service.managed-cluster](https://github.com/zedy-wj/bicep-registry-modules/actions/workflows/avm.res.container-service.managed-cluster.yml)|

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

@AlexanderSehr, @jongio for notification.

### CODE REVIEW

File: avm/res/container-service/managed-cluster/main.bicep
Comment (jongio) [inline avm/res/container-service/managed-cluster/main.bicep:134]: Do we really need this screaming caps error message in so many place?

Comment (zedy-wj) [reply avm/res/container-service/managed-cluster/main.bicep:134]: > Do we really need this screaming caps error message in so many place?

@jongio - This is required in the development documentation [spec](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr1---category-composition---preview-services).

@AlexanderSehr - Do you have any ideas?

Comment (AlexanderSehr) [reply avm/res/container-service/managed-cluster/main.bicep:134]: Hey @jongio,
this is kind of the idea - screaming at the user as it is a preview feature. It is certainly not pretty no - and truth be told I'm not the biggest fan myself. But the intent behind the spec makes sense. 
@matebarabas / @jtracey93 I guess there's not really a short handle, is there?

One small change I would suggest though @zedy-wj would be to change the text to `Optional. AUTO MODE IS A PARAMETER USED FOR A PREVIEW FEATURE, MICROSOFT MAY NOT PROVIDE SUPPORT FOR THIS, PLEASE CHECK THE [PRODUCT DOCS](https://learn.microsoft.com/en-us/azure/aks/learn/quick-kubernetes-automatic-deploy?pivots=bicep#before-you-begin) FOR CLARIFICATION. Node provisioning settings that apply to the whole cluster.`

Which would render the URL as a hyperlink.

Comment (zedy-wj) [reply avm/res/container-service/managed-cluster/main.bicep]: @AlexanderSehr - Updated.

Comment (matebarabas) [reply avm/res/container-service/managed-cluster/main.bicep]: @AlexanderSehr, I like the idea of adding the link to the related docs.

File: avm/res/container-service/managed-cluster/README.md
Comment (AlexanderSehr) [reply avm/res/container-service/managed-cluster/README.md:1]: Hey @zedy-wj,
do I see this correctly that a contributor would need to enable all of the following features?
```pwsh
az feature register --namespace 'Microsoft.ContainerService' --name 'EnableAPIServerVnetIntegrationPreview'
az feature register --namespace 'Microsoft.ContainerService' --name 'NRGLockdownPreview'
az feature register --namespace 'Microsoft.ContainerService' --name 'SafeguardsPreview'
az feature register --namespace 'Microsoft.ContainerService' --name 'NodeAutoProvisioningPreview'
az feature register --namespace 'Microsoft.ContainerService' --name 'DisableSSHPreview'
az feature register --namespace 'Microsoft.ContainerService' --name 'AutomaticSKUPreview'
```

Comment (zedy-wj) [reply avm/res/container-service/managed-cluster/README.md:1]: Hey @wbreza,
we are promoting `automatic cluster` related feature in AVM. Since some preview features require registering the feature flags, we are not very clear here. When adding related feature according to your [PR](https://github.com/Azure/azure-dev/pull/4069/files#diff-b13168826911b7f1ec80445aa91011c9a0e3430e1f4555cdb9276a141d877872), what I can confirm at present is that
```
az feature register --namespace Microsoft.ContainerService --name NodeAutoProvisioningPreview
az feature register --namespace Microsoft.ContainerService --name AutomaticSKUPreview
```
needs to be registered. Could you help confirm whether we need to register the remaining feature flags according to the [instructions of the doc](https://learn.microsoft.com/en-us/azure/aks/learn/quick-kubernetes-automatic-deploy?pivots=bicep#register-the-feature-flags)?
```
az feature register --namespace Microsoft.ContainerService --name EnableAPIServerVnetIntegrationPreview
az feature register --namespace Microsoft.ContainerService --name NRGLockdownPreview
az feature register --namespace Microsoft.ContainerService --name SafeguardsPreview
az feature register --namespace Microsoft.ContainerService --name DisableSSHPreview
```

Comment (v-xuto) [reply avm/res/container-service/managed-cluster/README.md:1]: @wbreza Any ideas about [above comments](https://github.com/Azure/bicep-registry-modules/pull/3646#discussion_r1825347139)?

Comment (AlexanderSehr) [inline avm/res/container-service/managed-cluster/README.md:1]: Please note, the `main.json` templates **must** be re-build with the latest Bicep version `0.31.34` as this version (running on the agents) renders a different `main.json` file then previous versions. 

The steps to execute would be
- Upgrade your Bicep CLI version to `0.31.34` (e.g. via `choco upgrade bicep` if choco is used)
- Re-run the `Set-AVMModule` script on your module. E.g. via `Set-AVMModule -ModuleFolderPath '.\avm\res\automation\automation-account\' -SkipFileAndFolderSetup -SkipReadMe -Recurse`
- Push the changes
- Re-run the tests to ensure everything works. 

If the static tests continue to fail for you, please note that there is currently presumably another new bug which affects some of our modules ([ref](https://github.com/Azure/bicep/issues/15517)).

Comment (zedy-wj) [reply avm/res/container-service/managed-cluster/README.md:1]: Updated.

**Comment by jtracey93** [comment]
Hey @zedy-wj ,

Firstly, thanks for your work on this PR!

We have made some changes to the AVM CI, detailed below, which means we need you to update your fork to pull in these latest changes and re-run your tests to show they still are passing prior to approving and merging this PR, as we don't and it fails once merged the publishing of your module will fail and will be blocked going forward until the test pass again via additional PRs.

### Changes to CI That Have Been Made That You Need To Take Action On

- https://github.com/Azure/bicep-registry-modules/pull/3745
    - Adds WAF Security PSRule Checks using this custom baseline of rules ([`/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml))
    - **Action Required By You:**
        - Ingest/pull-in these changes into your PR branch and re-run your tests to ensure they pass
            - If any issues with the rules and you think it should apply, please ping/reach out to the core team

Any questions reach out to the AVM Core Team by tagging us in your PR here or internally via Teams

Thanks

Jack (AVM Core Team)

**Comment by zedy-wj** [comment]
> Hey @zedy-wj ,
> 
> Firstly, thanks for your work on this PR!
> 
> We have made some changes to the AVM CI, detailed below, which means we need you to update your fork to pull in these latest changes and re-run your tests to show they still are passing prior to approving and merging this PR, as we don't and it fails once merged the publishing of your module will fail and will be blocked going forward until the test pass again via additional PRs.
> 
> ### Changes to CI That Have Been Made That You Need To Take Action On
> * [feat: Add WAF Security PS Rule Config #3745](https://github.com/Azure/bicep-registry-modules/pull/3745)
>   
>   * Adds WAF Security PSRule Checks using this custom baseline of rules ([`/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml))
>   * **Action Required By You:**
>     
>     * Ingest/pull-in these changes into your PR branch and re-run your tests to ensure they pass
>       
>       * If any issues with the rules and you think it should apply, please ping/reach out to the core team
> 
> Any questions reach out to the AVM Core Team by tagging us in your PR here or internally via Teams
> 
> Thanks
> 
> Jack (AVM Core Team)

Updated. The rerun tests have all passed.

File: avm/res/container-service/managed-cluster/README.md
Comment (AlexanderSehr) [reply avm/res/container-service/managed-cluster/README.md:1]: Has there been any progression here? The module has been blocked for quite a while - so much so that it may start to make more sense to remove the previous feature to re-enable updates...

File: avm/res/container-service/managed-cluster/main.bicep
Comment (AlexanderSehr) [reply avm/res/container-service/managed-cluster/main.bicep]: @jongio are we good?

**Review by JPEasier — APPROVED** [review]
looks good for me.

Thank you very much for your contribution!

**Review by JPEasier — DISMISSED** [review]

**Comment by v-xuto** [comment]
@rajeshkamal5050 Could you please contact Jongio and Wallace to review this PR? Here are the comments they left.

**Review by ilhaan — APPROVED** [review]

**Comment by AlexanderSehr** [comment]
For reference, [this](https://github.com/Azure/bicep-registry-modules/issues/3802) is why this PR is so important

File: avm/res/container-service/managed-cluster/main.bicep
Comment (rajeshkamal5050) [reply avm/res/container-service/managed-cluster/main.bicep]: @jongio can you review the updates above and unblock?

File: avm/res/container-service/managed-cluster/README.md
Comment (rajeshkamal5050) [reply avm/res/container-service/managed-cluster/README.md:1]: @wbreza can you help review the above doc/readme updates w.r.t to preview feature enablement?

**Comment by JPEasier** [comment]
@zedy-wj please resolve the conflicts be updating bicep cli and run the SetAvmModule script again.
Thanks a lot.

**Comment by ilhaan** [comment]
Closing this PR in favor of https://github.com/Azure/bicep-registry-modules/pull/3920

**Pull request closed without merging**
